arithmetic/src/multiplier/WallaceMultiplier.scala

```diff
@@ -106,49 +108,88 @@ class WallaceMultiplierImpl(
     }
     Seq.tabulate(pp.getWidth) { j => (i + j, pp(j)) } :+ (i, shouldPlus1)
   }
 
   val columnsMap = Booth
-    .recode(width)(radixLog2, signed = signed)(a.asUInt)
+    .recode(aWidth)(radixLog2, signed = signed)(a.asUInt)
     .zipWithIndex
     .flatMap { case (x, i) => makePartialProducts(radixLog2 * i, x) }
     .groupBy { _._1 }
 
-  val columns = Array.tabulate(2 * width) { i => columnsMap(i).map(_._2) }
+  val columns = Array.tabulate(aWidth + bWidth) { i => columnsMap(i).map(_._2) }
 
   val (sum, carry) = addAll(cols = columns, depth = 0)
 
-  val result = addition.prefixadder.apply(sumUpAdder)(sum, carry)(2 * width - 1, 0)
+  val result = addition.prefixadder.apply(sumUpAdder)(sum, carry)(aWidth + bWidth - 1, 0)
   z := (if (signed) result.asSInt else result.asUInt)
 }
 
 class SignedWallaceMultiplier(
-  val width:  Int
+  val aWidth: Int,
+  val bWidth: Int
 )(radixLog2:  Int = 2,
   sumUpAdder: PrefixSum = BrentKungSum,
   pipeAt:     Seq[Int] = Nil)
     extends SignedMultiplier {
-  val impl = Module(new WallaceMultiplierImpl(width, true)(radixLog2, sumUpAdder, pipeAt))
-  impl.a := a
-  impl.b := b
-  z := impl.z
+  if (aWidth < bWidth) {
+    val impl = Module(new WallaceMultiplierImpl(aWidth, bWidth, true)(radixLog2, sumUpAdder, pipeAt))
+    impl.a := a
+    impl.b := b
+    z := impl.z
+  } else {
+    val impl = Module(new WallaceMultiplierImpl(bWidth, aWidth, true)(radixLog2, sumUpAdder, pipeAt))
+    impl.a := b
+    impl.b := a
+    z := impl.z
+  }
 }
 
 class UnsignedWallaceMultiplier(
-  val width:  Int
+  val aWidth: Int,
+  val bWidth: Int
 )(radixLog2:  Int = 2,
   sumUpAdder: PrefixSum = BrentKungSum,
   pipeAt:     Seq[Int] = Nil)
     extends UnsignedMultiplier {
-  val impl = Module(new WallaceMultiplierImpl(width, false)(radixLog2, sumUpAdder, pipeAt))
-  impl.a := a
-  impl.b := b
-  z := impl.z
+  if (aWidth < bWidth) {
+    val impl = Module(new WallaceMultiplierImpl(aWidth, bWidth, true)(radixLog2, sumUpAdder, pipeAt))
+    impl.a := a
+    impl.b := b
+    z := impl.z
+  } else { // swap(a,b)
+    val impl = Module(new WallaceMultiplierImpl(bWidth, aWidth, true)(radixLog2, sumUpAdder, pipeAt))
+    impl.a := b
+    impl.b := a
+    z := impl.z
+  }
+}
+
+class FixedPointWallaceMultiplier(
+  val aWidth:   Int,
+  val aBPWidth: Int,
+  val bWidth:   Int,
+  val bBPWidth: Int
+)(radixLog2:    Int = 2,
+  sumUpAdder:   PrefixSum = BrentKungSum,
+  pipeAt:       Seq[Int] = Nil)
+    extends FixedPointMultiplier {
+  if (aWidth < bWidth) {
+    val impl = Module(new WallaceMultiplierImpl(aWidth, bWidth, true)(radixLog2, sumUpAdder, pipeAt))
+    impl.a := a.asSInt
+    impl.b := b.asSInt
+    z := impl.z.asFixedPoint((aBPWidth + bBPWidth).BP)
+  } else {
+    val impl = Module(new WallaceMultiplierImpl(bWidth, aWidth, true)(radixLog2, sumUpAdder, pipeAt))
+    impl.a := b.asSInt
+    impl.b := a.asSInt
+    z := impl.z.asFixedPoint((aBPWidth + bBPWidth).BP)
+  }
+
 }
```

Comment:
I think we should get rid of FixedPoint and even `SInt` and have a UInt version which can dynamic serve for Uint and SInt